File: server/pipeline/ocr/cover_timing.py
"""Cửa sổ che chữ — khớp LivePreviewEditor.coverWindow / burn cues.

Lead nhỏ ở đầu (tránh bbox/mask hiện trước chữ).
Tail / t_after: kéo dài che tới khi chữ thật sự biến mất.
"""
from __future__ import annotations

from typing import Any


def cover_lead_tail(layout: str, *, start: float, end: float, source: str = "") -> tuple[float, float]:
    """(lead, tail) giây nới quanh [start,end) theo layout."""
    lay = layout if layout in ("horizontal", "vertical", "label", "mid") else "horizontal"
    s0, e0 = float(start), float(end)
    dur = max(0.0, e0 - s0)
    if lay == "label":
        # OCR first-hit hay trễ hơn chữ — lead đủ che sớm
        return 0.28, 0.12
    if lay == "mid":
        return 0.35, 0.22
    if lay == "vertical":
        # watermark: OCR hay trễ nhẹ
        return (0.28 if dur >= 2.5 else 0.15, 0.0)
    src_cjk = sum(1 for c in (source or "") if "\u4e00" <= c <= "\u9fff")
    lead = 0.25  # hardsub ASR hay sớm hơn mực
    tail = 1.05 if dur <= 0.75 and src_cjk <= 4 else 0.45
    return lead, max(0.4, tail)


def default_cover_window(
    *,
    start: float,
    end: float,
    layout: str = "horizontal",
    source: str = "",
    video_end: float | None = None,
) -> tuple[float, float]:
    """Fallback cover khi segment chưa có coverStart/coverEnd."""
    s0, e0 = float(start), float(end)
    lead, tail = cover_lead_tail(layout, start=s0, end=e0, source=source)
    cs = max(0.0, s0 - lead)
    if layout == "vertical":
        ce = max(e0, cs + 0.04)
    elif layout == "label":
        ce = max(e0 + tail, cs + 0.20)
    elif layout == "mid":
        ce = max(e0 + tail, cs + 0.12)
    else:
        ce = max(e0 + tail, cs + 0.20)
    if video_end is not None:
        ce = min(float(video_end), ce)
    return round(cs, 3), round(max(ce, cs + 0.04), 3)


def resolve_cover_window(seg: dict[str, Any], *, video_end: float | None = None) -> tuple[float, float]:
    """Ưu tiên coverStart/coverEnd đã lưu; không thì default theo layout.

    Mid/label: nếu coverStart kéo sớm quá (OCR cũ → mốc trống trước) thì kẹp gần start.
    """
    s0 = float(seg.get("start") or 0)
    e0 = float(seg.get("end") or 0)
    layout = str(seg.get("layout") or "horizontal")
    source = str(seg.get("source") or "")
    cs_raw, ce_raw = seg.get("coverStart"), seg.get("coverEnd")
    if cs_raw is not None and ce_raw is not None:
        try:
            cs = float(cs_raw)
            ce = float(ce_raw)
            if ce > cs + 1e-6:
                # OCR cũ kéo coverStart về mốc trống quá sớm → kẹp, nhưng cho lead ~0.35 (OCR trễ)
                if layout in ("mid", "label") and cs < s0 - 0.45:
                    cs = max(0.0, s0 - 0.35)
                # ponytail: coverEnd lưu có thể ngắn hơn clip — luôn phủ hết [start,end)+tail
                _lead, tail = cover_lead_tail(layout, start=s0, end=e0, source=source)
                ce = max(ce, e0, cs + 0.04)
                if layout in ("mid", "label"):
                    ce = max(ce, e0 + tail)
                if video_end is not None:
                    ce = min(float(video_end), ce)
                return round(max(0.0, cs), 3), round(max(ce, cs + 0.04), 3)
        except (TypeError, ValueError):
            pass
    return default_cover_window(
        start=s0,
        end=float(seg.get("end") or 0),
        layout=layout,
        source=str(seg.get("source") or ""),
        video_end=video_end,
    )


def attach_cover_times(
    seg: dict[str, Any],
    *,
    t_before: float | None = None,
    t_after: float | None = None,
    video_end: float | None = None,
    neighbor_empty_before: bool = False,
    neighbor_empty_after: bool = False,
) -> dict[str, Any]:
    """Gán coverStart/coverEnd.

    - Đầu: lead trước start (OCR hay trễ); nếu mốc trước trống → kéo coverStart về gần t_before.
    - Cuối: khi mốc kế trống → kéo coverEnd tới t_after (+tail) để hết lộ chữ cũ.
    """
    s0 = float(seg.get("start") or 0)
    e0 = float(seg.get("end") or 0)
    layout = str(seg.get("layout") or "horizontal")
    source = str(seg.get("source") or "")
    _lead, tail = cover_lead_tail(layout, start=s0, end=e0, source=source)
    vend = float(video_end) if video_end is not None else None

    cs, ce = default_cover_window(
        start=s0, end=e0, layout=layout, source=source, video_end=vend
    )
    # Mid/label: giữ lead mặc định. Mốc trống trước → che sớm hơn first-hit (chữ đã có trong khe).
    if (
        layout in ("mid", "label", "vertical")
        and neighbor_empty_before
        and t_before is not None
        and float(t_before) < s0 - 0.05
    ):
        early = max(0.0, float(t_before) + 0.08)
        cs = min(cs, early)
        # không sớm hơn t_before; có thể sớm hơn start (đúng: chữ xuất trong khe)

    if neighbor_empty_after and t_after is not None and float(t_after) > e0 + 1e-3:
        # chữ còn tới ~mốc trống — kéo coverEnd tới đó + tail nhỏ
        ce = max(ce, float(t_after) + (0.0 if layout == "vertical" else min(tail, 0.22)))
        if vend is not None:
            ce = min(vend, ce)

    cs = max(0.0, round(cs, 3))
    ce = round(max(ce, cs + 0.04, e0), 3)
    if vend is not None:
        ce = min(vend, ce)
    seg["coverStart"] = cs
    seg["coverEnd"] = ce
    return seg

File: src/lib/coverTiming.ts
import type { Segment } from '../types'

/** Cửa sổ che chữ — khớp server/pipeline/ocr/cover_timing.py + burn cues.
 * Lead nhỏ ở đầu (tránh mask hiện trước chữ). Tail giữ dài hơn.
 */

export type CoverWindow = { start: number; end: number }

export function coverLeadTail(
  layout: string,
  start: number,
  end: number,
  source = '',
): { lead: number; tail: number } {
  const lay = layout || 'horizontal'
  const dur = Math.max(0, end - start)
  if (lay === 'vertical') {
    // watermark: OCR hay trễ nhẹ
    return { lead: dur >= 2.5 ? 0.28 : 0.15, tail: 0 }
  }
  if (lay === 'label') return { lead: 0.28, tail: 0.12 }
  if (lay === 'mid') return { lead: 0.35, tail: 0.22 }
  let srcCjk = 0
  for (const c of source) {
    if (c >= '\u4e00' && c <= '\u9fff') srcCjk += 1
  }
  const lead = 0.25
  const tail = end - start <= 0.75 && srcCjk <= 4 ? 1.05 : 0.45
  return { lead, tail: Math.max(0.4, tail) }
}

export function defaultCoverWindow(seg: Pick<Segment, 'start' | 'end' | 'layout' | 'source'>): CoverWindow {
  const s0 = seg.start
  const e0 = seg.end
  const layout = seg.layout || 'horizontal'
  const { lead, tail } = coverLeadTail(layout, s0, e0, seg.source || '')
  let start = Math.max(0, s0 - lead)
  // mid/label: cho lead ~0.35 (OCR first-hit hay trễ); không kẹp 80ms
  if (layout === 'mid' || layout === 'label') {
    start = Math.max(0, Math.min(start, s0 - 0.08))
  }
  if (layout === 'vertical') {
    return { start, end: Math.max(e0, start + 0.04) }
  }
  if (layout === 'label') {
    return { start, end: Math.max(e0 + tail, start + 0.2) }
  }
  if (layout === 'mid') {
    return { start, end: Math.max(e0 + tail, start + 0.12) }
  }
  return { start, end: Math.max(e0 + Math.max(0.4, tail), start + 0.2) }
}

/** Ưu tiên coverStart/coverEnd đã lưu; fallback default. */
export function resolveCoverWindow(seg: Segment): CoverWindow {
  const cs = seg.coverStart
  const ce = seg.coverEnd
  if (
    typeof cs === 'number' &&
    typeof ce === 'number' &&
    Number.isFinite(cs) &&
    Number.isFinite(ce) &&
    ce > cs + 1e-6
  ) {
    // segment đã lưu quá sớm (OCR cũ kéo về mốc trống) → kẹp mid/label
    const layout = seg.layout || 'horizontal'
    let start = Math.max(0, cs)
    if ((layout === 'mid' || layout === 'label') && start < seg.start - 0.45) {
      start = Math.max(0, seg.start - 0.35)
    }
    const { tail } = coverLeadTail(layout, seg.start, seg.end, seg.source || '')
    // ponytail: coverEnd lưu có thể ngắn hơn clip timeline — luôn phủ hết [start,end)
    let end = Math.max(ce, seg.end, start + 0.04)
    if (layout === 'mid' || layout === 'label') {
      end = Math.max(end, seg.end + tail)
    }
    return { start, end }
  }
  return defaultCoverWindow(seg)
}

/** True khi cửa sổ che khác rõ so với dịch (hiện dòng "Che …"). */
export function coverDiffersFromDub(seg: Segment): boolean {
  const w = resolveCoverWindow(seg)
  return Math.abs(w.start - seg.start) > 0.04 || Math.abs(w.end - seg.end) > 0.04
}
